File: src/org/jgll/sppf/TerminalSymbolNode.java
package org.jgll.sppf;

import java.util.Collections;
import java.util.Iterator;

import org.jgll.traversal.SPPFVisitor;

/**
 * 
 * 
 * @author Ali Afroozeh
 *
 */
public class TerminalSymbolNode extends SPPFNode {
	
	public static final int EPSILON = -2;
	
	private final int matchedChar;
	
	private final int inputIndex;
	
	public TerminalSymbolNode(int matchedChar, int inputIndex) {
		this.matchedChar = matchedChar;
		this.inputIndex = inputIndex;
	}
	
	@Override
	public boolean equals(Object obj) {
		
		if(this == obj) {
			return true;
		}

		if (!(obj instanceof TerminalSymbolNode)) {
			return false;
		}
		
		TerminalSymbolNode other = (TerminalSymbolNode) obj;
		
		return matchedChar == other.matchedChar &&
			   inputIndex == other.inputIndex;
	}
	
	@Override
	public int hashCode() {
		int result = 17;
		result += 31 * result + matchedChar;
		result += 31 * result + inputIndex;
		return result;
	}
	
	@Override
	public String getLabel() {
		return matchedChar == EPSILON ? "\u03B5" : (char) matchedChar + "";
	}
	
	@Override
	public String toString() {
		return String.format("(%s, %d, %d)", getLabel(), inputIndex, getRightExtent());
	}
	
	@Override
	public String getId() {
		return "t" + matchedChar + "," + inputIndex + "," + getRightExtent();
	}
	
	public int getMatchedChar() {
		return matchedChar;
	}
	
	@Override
	public int getLeftExtent() {
		return inputIndex;
	}

	@Override
	public int getRightExtent() {
		return matchedChar == EPSILON ? inputIndex : inputIndex + 1;
	}

	@Override
	public void accept(SPPFVisitor visitAction) {
		visitAction.visit(this);
	}

	@Override
	public Iterator<SPPFNode> iterator() {
		return Collections.<SPPFNode>emptyList().iterator();
	}

	@Override
	public SPPFNode get(int index) {
		return null;
	}

	@Override
	public int size() {
		return 0;
	}
}

File: src-scala/org/jgll/sppf/TerminalSymbolNode.scala
package org.jgll.sppf

import org.jgll.parser.HashFunctions
import org.jgll.traversal.SPPFVisitorTrait
import org.jgll.util.hashing.ExternalHasher
import org.jgll.util.hashing.hashfunction.HashFunction
import scala.reflect.BeanProperty
import org.jgll.grammar.slot.GrammarSlotTrait
import scala.collection.mutable.ListBuffer

//remove if not needed

trait TerminalSymbolNodeTrait {
  self: SPPFNodeTrait with SPPFVisitorTrait with GrammarSlotTrait =>
  import TerminalSymbolNode._
  object TerminalSymbolNode {

    val externalHasher = new TerminalSymbolNodeExternalHasher()

    val EPSILON = -2

    @SerialVersionUID(1L)
    class TerminalSymbolNodeExternalHasher extends ExternalHasher[TerminalSymbolNode] {

      override def hash(t: TerminalSymbolNode, f: HashFunction): Int = f.hash(t.inputIndex, t.matchedChar)

      override def equals(t1: TerminalSymbolNode, t2: TerminalSymbolNode): Boolean = {
        t1.matchedChar == t2.matchedChar && t1.inputIndex == t2.inputIndex
      }
    }
  }

  class TerminalSymbolNode(@BeanProperty val matchedChar: Int, private val inputIndex: Int)
    extends SPPFNode {

    override def equals(obj: Any): Boolean = {
      if (!(obj.isInstanceOf[TerminalSymbolNode])) {
        return false
      }
      val other = obj.asInstanceOf[TerminalSymbolNode]
      matchedChar == other.matchedChar && inputIndex == other.inputIndex
    }

    override def hashCode(): Int = {
      externalHasher.hash(this, HashFunctions.defaulFunction())
    }

    override def getLabel(): String = {
      if (matchedChar == EPSILON) "ε" else matchedChar.toChar + ""
    }

    override def toString(): String = {
      "(%s, %d, %d)".format(getLabel, inputIndex, getRightExtent)
    }

    override def getLeftExtent(): Int = inputIndex

    override def getRightExtent(): Int = {
      if (matchedChar == EPSILON) inputIndex else inputIndex + 1
    }

    override def accept(visitAction: SPPFVisitor) {
      visitAction.visit(this)
    }

    override def getChildAt(index: Int): SPPFNode = null

    override def childrenCount(): Int = 0

    override def getChildren(): Iterable[SPPFNode] = ListBuffer()

    override def isAmbiguous(): Boolean = false

    override def getGrammarSlot(): GrammarSlot = {
      throw new UnsupportedOperationException()
    }

    override def getLevel(): Int = inputIndex
  }
}